src/web/arr/trove/chart.arr

```diff
@@ -0,0 +1,1028 @@
+  values :: List<Number>,
+  # end for decoration
+  get-data :: ( -> Any)
+}
+
+default-pie-chart-series :: PieChartSeries  = default-series.{
+  labels: empty,
+  values: empty,
+}
+
+############
+
+type BarChartSeries = {
+  # for decoration
+  labels :: List<String>,
+  value-lists :: List<List<Number>>,
+  # end for decoration
+  get-data :: ( -> Any)
+}
+
+default-bar-chart-series :: BarChartSeries = default-series.{
+  labels: empty,
+  value-lists: empty,
+}
+
+############
+
+type HistogramSeries = {
+  # for decoration
+  labels :: List<String>,
+  values :: List<Number>,
+  # end for decoration
+  bin-width :: Number,
+  max-num-bins :: Number,
+  min-num-bins :: Number,
+  value-name :: String,
+  get-data :: ( -> Any)
+}
+
+default-histogram-series :: HistogramSeries  = default-series.{
+  labels: empty,
+  values: empty,
+  bin-width: 0,
+  max-num-bins: 0,
+  min-num-bins: 0,
+  value-name: 'Value',
+}
+
+############
+
+type LinePlotSeries = {
+  # for decoration
+  xs :: List<Number>,
+  ys :: List<Number>,
+  # end for decoration
+  color :: I.Color,
+  legend :: String,
+  get-data :: ( -> Any)
+}
+
+default-line-plot-series :: LinePlotSeries = default-series.{
+  xs: empty,
+  ys: empty,
+  color: I.red,
+  legend: '',
+}
+
+############
+
+type ScatterPlotSeries = {
+  # for decoration
+  xs :: List<Number>,
+  ys :: List<Number>,
+  # end for decoration
+  labels :: List<String>,
+  color :: I.Color,
+  legend :: String,
+  point-size :: Number,
+  get-data :: ( -> Any)
+}
+
+default-scatter-plot-series :: ScatterPlotSeries = default-series.{
+  xs: empty,
+  ys: empty,
+  labels: empty,
+  color: I.blue,
+  legend: '',
+  point-size: 7
+}
+
+############
+
+type FunctionPlotSeries = {
+  f :: PlottableFunction,
+  color :: I.Color,
+  legend :: String,
+  get-data :: ( -> Any)
+}
+
+default-function-plot-series :: FunctionPlotSeries = default-series.{
+  f: {(x): x},
+  color: I.blue,
+  legend: '',
+}
+
+###########
+
+type WindowOptions = {
+  x-min :: Number,
+  x-max :: Number,
+  y-min :: Number,
+  y-max :: Number,
+  num-samples :: Number
+}
+
+type ChartWindowObject = {
+  interact :: Boolean,
+  title :: String,
+  width :: Number,
+  height :: Number,
+  x-axis :: String,
+  y-axis :: String,
+  x-min :: Number,
+  x-max :: Number,
+  y-min :: Number,
+  y-max :: Number,
+  num-samples :: Number,
+  render :: ( -> IM.Image)
+}
+
+default-chart-window-object :: ChartWindowObject = {
+  interact: true,
+  title: '',
+  width: 400,
+  height: 300,
+  x-axis: '',
+  y-axis: '',
+  x-min: 0,
+  x-max: 0,
+  y-min: 0,
+  y-max: 0,
+  num-samples: 1000,
+  method render(self): raise('unimplemented') end,
+}
+
+################################################################################
+# DATA DEFINITIONS
+################################################################################
+
+data DataSeries:
+  | line-plot-series(obj :: LinePlotSeries) with:
+    is-single: false,
+    constr: {(): line-plot-series},
```

Comment:
because without it being thunk, they are unbound-id.

Reply:
That's an interesting design point for how `data` blocks desugar. Right now the values in the `with:` block are letrec-bound before the constructor is created, so you get a dynamic “this name not initialized yet”; this is an example where it would be better if that weren't true. But that would require an extra bit of state to create the constructor, then create its `with:` fields, then tie the loop cycle later to make sure the constructed objects can refer to these fields.

@justinpombrio @shriram this is an interesting scope desugaring case. I don't recall making an explicit design decision here, and I think this just fell out of the “obvious” implementation.

Reply:
I'm pretty sure this has bitten us in the past before.  It's challenging, since we want to have a shared `base` object for every instance of a given constructor, but I think we need to split the fields that are methods from fields that aren't (which are syntactically easily distinguishable), and only the methods go on the shared base object.  Then, I think a safe but ugly desugaring for this would be
```js
function makeConstr(arg1, arg2, arg3) {
  var dict = runtime.create(baseConstr);
  dict["arg1Name"] = arg1; dict["arg2Name"] = arg2; dict["arg3Name"] = arg3;
  var ans = new Construct(dict, brands);
  dict["withField1"] = computeWithField1(ans); // gotta thunk it
  return ans;
}
```